pallets/parachain-staking/src/lib.rs

```diff
@@ -437,30 +436,86 @@ pub mod pallet {
 		}
 	}
 
+	#[derive(Clone, PartialEq, Encode, Decode, RuntimeDebug)]
+	pub enum NominatorStatus {
+		/// Active with no scheduled exit
+		Active,
+		/// Schedule exit to revoke all ongoing nominations
+		Leaving(RoundIndex),
+	}
+
+	#[derive(Encode, Decode, RuntimeDebug)]
+	/// Nominator state
+	pub struct Nominator2<AccountId, Balance> {
```

Comment:
It would be awesome if we could architect this migration such that the pallet itself just changes the `Nominator` struct (in a non-backwards-compatible way) and we put all the gory details of what the original struct was and how to migrate in a separate file/lib/etc.

This is still a bit vague to me, but I'd be happy to help figure out what this should look like you want to go down this path.

Reply:
See the last sentence of the PR description:

```
From storage map NominatorState -> NominatorState2 such that the keys (AccountId) stay the same, but the value for the map changes from struct Nominator -> struct Nominator2. **There is an implementation of From<Nominator> for Nominator2 which is what is used to migrate the storage values from the first map to the second.**
```

I can show you where this is. You can search `impl From<Nominator<AccountId, Balance>> for Nominator2<AccountId, Balance>` and it will show you where this logic is.

Reply:
Ah, I see. `From` is an elegant way to handle that.

I guess, then, my suggestion would come down to:

1) Rename `Nominator` -> `OldNominator` (etc)
2) Move it to a different file (for clarity and/or potentially being excluded from future runtimes)
3) Rename `Nominator2` -> `Nominator`

Reply:
It looks like pulling a `#[pallet::storage]` definition out of the main pallet isn't really possible, but these utilities might help:

https://crates.parity.io/frame_support/storage/migration/index.html

I'll play with this to see if I can come up with something clean.

Reply:
> Ah, I see. `From` is an elegant way to handle that.
> 
> I guess, then, my suggestion would come down to:
> 
> 1. Rename `Nominator` -> `OldNominator` (etc)
> 2. Move it to a different file (for clarity and/or potentially being excluded from future runtimes)
> 3. Rename `Nominator2` -> `Nominator`

I had the same initial feeling when we did the previous migration which moved from CollatorState to CollatorState2. I had 2 reasons for that:
1. It is more elegant in the code to not have a number at the end of a variable (not too important)
2. It changes the way we query for the CollatorState (chain state -> parachainStaking -> CollatorState2)

But after thinking more about it, I'm not sure if it makes more sense to try to keep a same (exposed) variable with a different type. Because in both cases (changing the name of the storage, or changing its type), it is a breaking change that requires the developer to update its tools/sdk. 
I'm open to discussion on this

Reply:
One idea for handling these cases:

The new version of the pallet depends on the old version. The old type definition and storage definition are imported from the old pallet.